File: Java/0001-0050/3. Longest Substring Without Repeating Characters.java
// https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

class Solution {
    public int lengthOfLongestSubstring(String s) {
        //use two pointer approach
        int start  = 0; 
        int end = 0;
        int max = 0;

        //HashSet to store unique characters
        HashSet<Character> storage = new HashSet<>();
        int length = 0; //length of current substring

        while(end < s.length()){
            char letter = s.charAt(end);
            while(storage.contains(letter)){
                storage.remove(s.charAt(start));
                length--;start++;
            }

            if(!storage.contains(letter)){
                //add to HashSet
                storage.add(letter);
                end++;length++;
                max = Math.max(max , length);
            }
        }
        return max;
    }
}

File: Python/0001-0050/3. Longest Substring Without Repeating Characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # use two pointer approach
        start  = 0 
        end = 0
        maxy = 0

        # HashSet to store unique characters
        storage = {"?/?/"}
        length = 0 # length of current substring

        while end < len(s):
            letter = s[end]
            while letter in storage:
                storage.remove(s[start])
                length -= 1
                start += 1
            

            if letter in storage:
                continue
            else:
                # add to HashSet
                storage.add(letter)
                end += 1
                length += 1
                maxy = max(maxy , length)
            
        
        return maxy
        
